### PR TITLE
[ENT-365] Move from old python-social-auth to latest, split PSA

### PR DESCRIPTION
**Pull request by UmanShahzad**
**Description:** Old [python-social-auth](https://github.com/python-social-auth) was a monolithic repository and the version we utilize is almost a year old. We are looking to upgrade to the latest, split version of python-social-auth which is more modular, and provides several new features, including a configurable maximum session length, a DB-packed partial pipeline, utilizing the latest backend (provider) versions, and so on.

**JIRA:** Implements [ENT-365](https://openedx.atlassian.net/browse/ENT-365).

**Dependencies:** [Complementary PR at edx-platform](https://github.com/edx/edx-platform/pull/15135).

**Merge deadline:** Mostly handles technical debt, so not that urgent.

**Installation instructions:** N/A for now.

**Testing instructions:** See [edx-platform PR](https://github.com/edx/edx-platform/pull/15135) for now.

**Merge checklist:**

- [ ] Regenerate requirements with `make upgrade && make requirements` (and make sure to fix any errors).
  **DO NOT** just add dependencies to `requirements/*.txt` files.
- [x] All reviewers approved
- [x] CI build is green
- [x] Version bumped
- [x] Changelog record added
- [x] Documentation updated (not only docstrings)
- [x] Commits are (reasonably) squashed
- [x] Translations are updated
- [x] PR author is listed in AUTHORS

**Post merge:**
- [ ] Create a tag
- [ ] Check new version is pushed to PyPi after tag-triggered build is finished.
- [ ] Delete working branch (if not needed anymore)

**Author concerns:** This **must** be merged at the **same time** as the corresponding [edx-platform PR](https://github.com/edx/edx-platform/pull/15135), or else things will break.

### CODE REVIEW

**Comment by edx-webhook** [comment]
Thanks for the pull request, @UmanShahzad! It looks like you're a member of a company that does contract work for edX. If you're doing this work as part of a paid contract with edX, you should talk to edX about who will review this pull request. If this work is not part of a paid contract with edX, then you should ensure that there is an OSPR issue to track this work in [JIRA](https://openedx.atlassian.net), so that we don't lose track of your pull request.

<a   href="https://openedx-webhooks.herokuapp.com/github/process_pr?number=124&repo=edx%2Fedx-enterprise"   rel="nofollow" >Create an OSPR issue for this pull request</a>.

File: tests/test_views.py [hunk 1]
```diff
@@ -202,26 +197,65 @@ def test_get_render_patched(
...
+        )
+        expected_context = {
+            'consent_request_prompt': expected_prompt,
+            'confirmation_alert_prompt': expected_alert,
+            'confirmation_alert_prompt_warning': expected_warning,
+            'platform_name': 'This Platform',
+            'enterprise_customer_name': 'Fake Customer Name',
+        }
+        for key, value in expected_context.items():
+            assert response.context[key] == value  # pylint: disable=no-member
```
Comment (UmanShahzad) [inline tests/test_views.py]: The above function is the only significant change in this file -- the rest is just adding some necessary mock patches and refactoring to make the file a little shorter and more consistent.

**Comment by UmanShahzad** [comment]
@douglashall and @brittneyexline -- this PR, along with https://github.com/edx/edx-platform/pull/15135, is ready for review. Note that we'll have to merge this one first, and at the same time update the edx-platform PR to use this latest edx-enterprise version, and then merge the latter.

**Comment by UmanShahzad** [comment]
@brittneyexline @douglashall Just a reminder -- we've got the edx-platform PR ready to go (upstream PSA PR is merged!), so only this needs review and the former needs approval, then we're good to go. Some OpenCraft dev will also be looking at this shortly.

File: enterprise/tpa_pipeline.py [hunk 2]
```diff
@@ -10,41 +10,58 @@
...
+    Ensure that all necessary third_party_auth dependencies are present.
+    """
+    third_party_dependencies = (
+        Registry,
+        get_partial_pipeline,
+    )
+
+    if any(third_party_dependency is None for third_party_dependency in third_party_dependencies):
+        raise NotConnectedToOpenEdX(
+            _("This package must be installed in an Open edX environment to look up third-party auth dependencies.")
```
Comment (brittneyexline) [inline enterprise/tpa_pipeline.py]: Maybe this message should be updated to say what package needs to be installed?

File: enterprise/utils.py [hunk 3]
```diff
@@ -49,19 +49,7 @@ def __init__(self, *args, **kwargs):
...
-        super(NotConnectedToEdX, self).__init__(*args, **kwargs)
-
-
-class NotConnectedToOpenEdX(Exception):
-    """
-    Exception to raise when we weren't able to import needed resources.
-
-    This is raised when we try to use the resources, rather than on
-    import, so that during testing we can load the module
-    and mock out the resources we need.
```
Comment (brittneyexline) [inline enterprise/utils.py]: Just curious about why we decided to consolidate these exceptions now? Not that I'm not in favor - I've never fully understood the difference between them.

File: enterprise/views.py [hunk 4]
```diff
@@ -74,10 +76,11 @@ def verify_edx_resources():
         get_enterprise_customer_for_request,
         get_real_social_auth_object,
         quarantine_session,
-        lift_quarantine
+        lift_quarantine,
+        get_partial_pipeline,
     )
     if any(method is None for method in required_methods):
-        raise NotConnectedToEdX(_('Methods in the Open edX platform necessary for this view are not available.'))
+        raise NotConnectedToOpenEdX(_('Methods in the Open edX platform necessary for this view are not available.'))
```
Comment (brittneyexline) [inline enterprise/views.py]: also worth updating the message here to be more specific about the missing dependencies.

File: enterprise/tpa_pipeline.py (hunk 2 above)
Comment (UmanShahzad) [inline enterprise/tpa_pipeline.py]: Sure! :+1:

File: enterprise/utils.py (hunk 3 above)
Comment (UmanShahzad) [inline enterprise/utils.py]: To tell you the truth, this PR was picked up from @haikuginger, and this change was a part of it, so I decided to keep it going. It doesn't look like to me the changes are exactly that significant, it's just a squashing of two exception classes to maybe simplify what it is we work with.

I guess its original form could be considered a sort of minor tech debt, and it would have been better to eventually clean it up than to simply ignore it up until a dedicated PR. Since it's here and not really bothering anyone, I suppose we can continue with it, is that okay? If not, I'll open a dedicated PR to do it but will take more work. Either that or I can separate the commits for this change and for the rest in this PR.

File: enterprise/views.py (hunk 4 above)
Comment (UmanShahzad) [inline enterprise/views.py]: Yep, easy peasy. :+1:

File: enterprise/utils.py (hunk 3 above)
Comment (brittneyexline) [inline enterprise/utils.py]: Yea, I'm in favor of keeping it in this PR, I just wanted to call it out since it wasn't essential to the purpose of the PR.

**Comment by UmanShahzad** [comment]
@brittneyexline Ok, I had to jump some hoops to get it to display the missing method/dependency name properly, including dealing with tests looking explicitly for matching error messages ( :| ) and a significant rebase, but it's done now.

@haikuginger should, hopefully, be able to review this soon.

When approved I want to squash this.

File: enterprise/views.py [hunk 5]
```diff
@@ -75,15 +77,21 @@ def verify_edx_resources():
...
-    if any(method is None for method in required_methods):
-        raise NotConnectedToEdX(_('Methods in the Open edX platform necessary for this view are not available.'))
+    required_methods = {
+        'configuration_helpers': configuration_helpers,
+        'get_enterprise_customer_for_request': get_enterprise_customer_for_request,
+        'get_real_social_auth_object': get_real_social_auth_object,
+        'quarantine_session': quarantine_session,
+        'lift_quarantine': lift_quarantine,
+        'get_partial_pipeline': get_partial_pipeline,
+    }
```
Comment (bradenmacdonald) [inline enterprise/views.py]: Nit (which means this is optional and you can ignore): instead of duplicating all the imports here, you could just change the code at the top of this file to say

```python
import_error = None

try:
    from ... import ...
except ImportError as e:
    import_error = import_error or e.message
```

Then you can just display `import_error` in the error message below:
```python
def verify_edx_resources():
    if import_error:
        raise NotConnectedToOpenEdX("An Open edX dependency required for this method is missing: {}".format(import_error))
```

Or another option is to just log a warning message within the `except ImportError` block itself, and don't bother calling `verify_edx_resources()` elsewhere.

File: enterprise/tpa_pipeline.py [hunk 6]
```diff
@@ -78,7 +98,10 @@ def active_provider_enforces_data_sharing(request, enforcement_location):
         enforcement_location (str): the point where to see data sharing consent state.
         argument can either be "optional", 'at_login' or 'at_enrollment'
     """
-    running_pipeline = request.session.get('partial_pipeline')
+    try:
+        running_pipeline = get_partial_pipeline(request)
+    except TypeError:
+        running_pipeline = request.session.get('partial_pipeline_token')
```
Comment (bradenmacdonald) [inline enterprise/tpa_pipeline.py]: What are the circumstances where `get_partial_pipeline()` does not work but `running_pipeline = request.session.get('partial_pipeline_token')` does? And aren't those different types? (`get_partial_pipeline` returns a `dict` but `request.session.get('partial_pipeline_token')` returns a `str`)

If this is for a test case, it should be mocked out or replaced with a fake (install a simple `third_party_auth` fake app in the test environment, full of `Mock`s), rather than being here in the regular code.

Comment (UmanShahzad) [inline enterprise/tpa_pipeline.py]: @bradenmacdonald I think this is to protect against the case when we're "NotConnectedToOpenEdX". I'm not sure if it's better to check for the dependency and remove the try-catch block, or to keep it going like this.

File: enterprise/views.py (hunk 5 above)
Comment (UmanShahzad) [inline enterprise/views.py]: @bradenmacdonald I actually agree with the above solutions, but I've been beating this exception bush quite a bit in this PR and I think I want to just dedicate a new, smaller PR with a cleaner approach to this; for now it'll do, if that's alright -- but definitely want to change it.

File: enterprise/tpa_pipeline.py (hunk 6 above)
Comment (bradenmacdonald) [inline enterprise/tpa_pipeline.py]: Hmm, generally code specific to tests should be kept in the `_test` files as much as possible, so that it's clear what is normal/expected behavior and what are the mocks/hacks/workarounds put in just for tests, which can be removed/refactored anytime.

In this case, can we just change the test code to mock out `get_partial_pipeline` rather than having this `try`/`except`?

File: enterprise/views.py (hunk 5 above)
Comment (bradenmacdonald) [inline enterprise/views.py]: Makes sense - fine by me to keep that for another PR!

File: enterprise/tpa_pipeline.py (hunk 6 above)
Comment (UmanShahzad) [inline enterprise/tpa_pipeline.py]: @bradenmacdonald The issue is actually that the test code where this fails _is_ mocking out `get_partial_pipeline`.

Do you think it's okay to just leave it as `request.session.get('partial_pipeline_token')`? In this function we're just checking for whether the pipeline is running and nothing more, so what type is returned doesn't really matter as long as it indicates to us that the pipeline is indeed running, which the token should be enough for?

**Comment by UmanShahzad** [comment]
@bradenmacdonald I gave a reply on the try-catch block issue. Not sure if this'll be enough, but let me know.

**Comment by haikuginger** [comment]
@UmanShahzad, I ran into issues while testing this - some in the edx-platform PR, some here. Specifically, I had issues with the quarantine stuff dropping my session after submitting data sharing consent. LMK if you need more details on that.

File: enterprise/tpa_pipeline.py (hunk 6 above)
Comment (bradenmacdonald) [inline enterprise/tpa_pipeline.py]: Yeah that makes sense to me.

**Comment by UmanShahzad** [comment]
@haikuginger Any idea of how to reproduce the error?

File: enterprise/tpa_pipeline.py [hunk 7]
```diff
@@ -78,7 +98,7 @@ def active_provider_enforces_data_sharing(request, enforcement_location):
         enforcement_location (str): the point where to see data sharing consent state.
         argument can either be "optional", 'at_login' or 'at_enrollment'
     """
-    running_pipeline = request.session.get('partial_pipeline')
+    running_pipeline = request.session.get('partial_pipeline_token')
```
Comment (UmanShahzad) [inline enterprise/tpa_pipeline.py]: @haikuginger I would say it's sufficient: we use the token to get the pipeline anyway, and getting the pipeline is thus wholly dependent on that. If the token doesn't exist, we wouldn't have gotten the pipeline anyway.

**Comment by UmanShahzad** [comment]
@haikuginger Thanks! Good to know the bug wasn't related. I answered your review question, hopefully it's sufficient. I'll test on Android, and if all's good (it should be since iOS/Android are related in this respect as @bradenmacdonald mentioned), I'll fix these branch conflicts and give you a ping.

**Comment by haikuginger** [comment]
@UmanShahzad, can you rebase this and update the version to 0.36.0? This is definitely worthy of a bump, and it looks like 0.35.2 was recently merged.

**Comment by UmanShahzad** [comment]
@haikuginger Done, and rebased.
